pkg/services/ngalert/state/manager.go

```diff
@@ -83,13 +89,18 @@ func NewManager(cfg ManagerCfg) *Manager {
 		maxStateSaveConcurrency:        cfg.MaxStateSaveConcurrency,
 		applyNoDataAndErrorToAllStates: cfg.ApplyNoDataAndErrorToAllStates,
 		tracer:                         cfg.Tracer,
+		stateRunnerShutdown:            make(chan interface{}, 1),
+		saveStateAsync:                 cfg.SaveStateAsync,
 	}
 }
 
 func (st *Manager) Run(ctx context.Context) error {
 	if st.applyNoDataAndErrorToAllStates {
 		st.log.Info("Running in alternative execution of Error/NoData mode")
 	}
+	if st.saveStateAsync {
+		st.startSync(ctx)
+	}
```

Comment:
I think this code could benefit from an abstraction - more than anything to make sure we exercise proper tests on them. 

I understand that the state manager is kind of hodgepodge at the moment, but adding more fire to the would not align with the [boyscout rule](https://levelup.gitconnected.com/the-boy-scouting-rule-a-simple-yet-effective-programming-principle-5d504c7dd993#:~:text=The%20Boy%20Scouting%20Rule%20is,environment%20cleaner%20than%20they%20found.).

I would love to see an abstraction over the saving of states; my suggestion would be something like `Persister` or `Persistence` (e.g. placed under `state/persister.go`) that allows you to swap out the implementation depending on the feature flag.


With that, your `NewManager` functions can end up being like:

```golang
func NewManager(cfg ManagerCfg) *Manager {
	m := &Manager{
		cache:                          newCache(),
		ResendDelay:                    ResendDelay, // TODO: make this configurable
		log:                            cfg.Log,
		metrics:                        cfg.Metrics,
		instanceStore:                  cfg.InstanceStore,
		images:                         cfg.Images,
		historian:                      cfg.Historian,
		clock:                          cfg.Clock,
		externalURL:                    cfg.ExternalURL,
		doNotSaveNormalState:           cfg.DoNotSaveNormalState,
		maxStateSaveConcurrency:        cfg.MaxStateSaveConcurrency,
		applyNoDataAndErrorToAllStates: cfg.ApplyNoDataAndErrorToAllStates,
		tracer:                         cfg.Tracer,
		stateRunnerShutdown:            make(chan interface{}, 1),
		persister:                      NewSyncPersister(),
	}

	if cfg.SaveStateAsync {
		m.persister = NewAsyncPersister()
	}

	return m
}
```

I recognise this is hard - we already have `persist.go` in this package but it contains mostly interfaces because the persistence layer uses the database to save some of these states. Another alternative for all of this is to create a subpackage called `recovery` (at the same level as historian for example) and then place everything related to this work there.

Reply:
I don't see which interface the sync and async approach would share, could you elaborate on this?

Reply:
They don't necessarily need to share a method - I thought they could have two methods and one would be no-ops in the other.

Example:

```golang
type SyncPersister struct {
}

// this operation no-ops in sync.
func (s *SyncPersister) ProcessAsync() {
	return
}

func (s *SyncPersister) Process() {
	st.deleteAlertStates(tracingCtx, logger, staleStates)
	if len(staleStates) > 0 {
		span.AddEvents([]string{"message", "state_transitions"},
			[]tracing.EventValue{
				{Str: "deleted stale states"},
				{Num: int64(len(staleStates))},
			})
	}

	st.saveAlertStates(tracingCtx, logger, states...)
	span.AddEvents([]string{"message"},
		[]tracing.EventValue{
			{Str: "updated database"},
		})

}

type AsyncPersister struct {
}

func (a *AsyncPersister) ProcessAsync() {
	instances := st.cache.asInstances(st.doNotSaveNormalState)
	// save them to SQL after
}

func (a *AsyncPersister) Process() {
	return
}
```

This is some pseudo-code as it doesn't compile -- if this is not clear enough, I'm happy to give you a git diff with working code if it helps.

Reply:
I guess the most important thing for me is that both of these are tested correctly in code.

Reply:
Gotcha, thanks for looping back.

Reply:
Do you need two methods on the interface? For example:

```go
type SyncPersister struct {
}

func (s *SyncPersister) Process() {
	st.deleteAlertStates(tracingCtx, logger, staleStates)
	if len(staleStates) > 0 {
		span.AddEvents([]string{"message", "state_transitions"},
			[]tracing.EventValue{
				{Str: "deleted stale states"},
				{Num: int64(len(staleStates))},
			})
	}

	st.saveAlertStates(tracingCtx, logger, states...)
	span.AddEvents([]string{"message"},
		[]tracing.EventValue{
			{Str: "updated database"},
		})

}

type AsyncPersister struct {
}

func (a *AsyncPersister) Process() {
	instances := st.cache.asInstances(st.doNotSaveNormalState)
	// save them to SQL after
}
```